### ISSUE TITLE
Removal of intree storage plugin is causing us to lose important test coverage

### ISSUE THREAD

**Issue by gnufied** [kubernetes/kubernetes#118360]
### What happened?

When we removed the intree plugins - https://github.com/kubernetes/kubernetes/pull/116301 

We are also removing the e2es and in-process losing coverage such as for pre-provisioned or inline volumes. This causes those code paths to be error prone.

IMO - my naive thinking was to remove the plugin actively from `kubernetes/pkg` etc, but at the same time keep the vendored dependencies so as we can build the e2e and keep using it.

### What did you expect to happen?

Run e2e and I would expect inline, pre-provisioned volume tests to still be running.

### How can we reproduce it (as minimally and precisely as possible)?

Run e2e and

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by gnufied** [comment]
/sig storage
/sig cloudprovider

cc @dims @cheftako @kubernetes/sig-storage-api-reviews

**Comment by msau42** [comment]
This approach will cover the test cases that are part of the storage framework, but not the test cases from pre-framework days: https://github.com/kubernetes/kubernetes/issues/88055

**Comment by dims** [comment]
> It is extremely desirable to eliminate the cloud SDKs from our vendor. They're really large dependencies.

+100000000000

**Comment by gnufied** [comment]
> It is extremely desirable to eliminate the cloud SDKs from our vendor. They're really large dependencies.

I agree. I remember there was some effort to move e2e out of main `k/k` repo, I wonder if it went anywhere.

> However, I think there is a way we can potentially retain some coverage (although may not be as extensive as what we used to have) without requiring cloud provider dependencies. 

Yeah I think we will have to experiment and come up with a way of making them work for now, otherwise we are losing some valuable coverage which is still fully supported.

**Comment by jsafrane** [comment]
It would be great to run these test, anyone is more than welcome to contribute. See comment https://github.com/kubernetes/kubernetes/issues/118360#issuecomment-1570750797 for a very high level idea how to proceed.

/triage-accepted
/help

**Comment by HirazawaUi** [comment]
I'd like work for this issue
/assign

**Comment by HirazawaUi** [comment]
@msau42
Apologies for the delay in my response.

After reviewing the code, I would like to clarify a few questions that have been causing confusion:
  
In the PR at https://github.com/kubernetes/kubernetes/pull/116301, the code related to e2e was not removed (it was initially removed but later restored). Could you please provide more specific details about the test coverage we lost?

The hooks in [CreateVolume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1153-L1170) seem to depend on the cloud provider. However, is it not our intention to eliminate the dependency on the cloud provider?

It appears that there are no dependencies on the volumes created by CreateVolume in [pre-provisioned volume snapshot tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/framework/snapshot_resource.go#L94-L99). Could you please provide further clarification?

I apologize for the multiple questions, and i mistakenly underestimated his difficulty in the beginning.

**Comment by gnufied** [comment]
Where do you see the e2e portion of code being restored?

> The hooks in [CreateVolume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1153-L1170) seem to depend on the cloud provider. However, is it not our intention to eliminate the dependency on the cloud provider?

Yes that is the problem in a nutshell.  Pre-Provisioned/Inline tests depend on the cloudprovider and we are working on removing cloudprovider code from in-tree and hence all tests that use pre-provisioned/inline volumes won't work.

It is not a easy problem to solve though. :(

**Comment by HirazawaUi** [comment]
> Where do you see the e2e portion of code being restored?

I see that in the commit [remove Azure Disk in-tree driver code](https://github.com/kubernetes/kubernetes/pull/116301/commits/5d0a54dcb5ab5a36363e853894a254a200c2ef81) and [fix intree test](https://github.com/kubernetes/kubernetes/pull/116301/commits/0b61e0d9f7fed54cae5487cc95310325ff3d7592) some codes in the e2e directory are removed, and in the [revert azuredisk test removal change](https://github.com/kubernetes/kubernetes/pull/116301/commits/c2b2a7622f0f51d102098119469d49f7544f21cd) they are restored

> Yes that is the problem in a nutshell. Pre-Provisioned/Inline tests depend on the cloudprovider and we are working on removing cloudprovider code from in-tree and hence all tests that use pre-provisioned/inline volumes won't work.

According to this, if I understand correctly, we need to modify the code in [CreateVolume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1153-L1170) in a way that eliminates the dependence on the cloud provider, is that right?

**Comment by gnufied** [comment]
Well according to https://github.com/kubernetes/kubernetes/pull/116301 , no matter the revert that code will be removed in 1.28. The PR itself implies that.

 > According to this, if I understand correctly, we need to modify the code in [CreateVolume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1153-L1170) in a way that eliminates the dependence on the cloud provider, is that right?

Yeah that is the idea.

**Comment by HirazawaUi** [comment]
> Well according to #116301 , no matter the revert that code will be removed in 1.28. The PR itself implies that.
> 
> > According to this, if I understand correctly, we need to modify the code in [CreateVolume](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1153-L1170) in a way that eliminates the dependence on the cloud provider, is that right?
> 
> Yeah that is the idea.

Thank you for the clarification

**Comment by dims** [comment]
Also see this comment (screenshotting so we don't lose it in the github comment shuffles)

https://github.com/kubernetes/kubernetes/pull/122966#discussion_r1466495795

<img width="1211" alt="image" src="https://github.com/kubernetes/kubernetes/assets/23304/4f38a6d3-2454-4a36-ae13-6474a4ba4293">

**Comment by HirazawaUi** [comment]
/unassign 

Because: https://github.com/kubernetes/kubernetes/pull/119926#issuecomment-2227394968
